### PR TITLE
Issue #4572 - Implementing old logging TAG_PAD as Message Alignment.

### LINKED ISSUE #4572: Replace Jetty Logging
# Proposal

## Drop org.eclipse.jetty.util.log

The entire package `org.eclipse.jetty.util.log` is deleted

## Create jetty-slf4j-impl module

A new module called `jetty-slf4j-impl` is created.

This module performs the equivalent to Jetty's StdErrLog.

* Same output formatting as `StdErrLog` in 9.x
* Supports `jetty-logging.properties` for configuring logging levels

It has a `META-INF/services/org.slf4j.spi.SLF4JServiceProvider`

A package space of org.eclipse.jetty.logging
A JPMS of jetty-logging ?

This new module would be included as a `<scope>test</scope>`
across the other Jetty modules that need it for testing.

## Migrate logging properties

Existing properties

* `org.eclipse.jetty.util.log.class` (classname, default "org.eclipse.jetty.util.log.Slf4jLog") - use to select the logger
* `org.eclipse.jetty.util.log.announce` (boolean, default "true") - used to control announcement text `"Logging to {} via {}"`
* `org.eclipse.jetty.util.log.IGNORED` (boolean, default "false") - used to control if Log.ignored() should be shown or not
* `org.eclipse.jetty.util.log.StdErrLog.TAG_PAD` (number, default "0") - used for minimum padding width of the Thread.name
* `org.eclipse.jetty.util.log.SOURCE` (boolean, default "false") - used as default for showing of source (filename and line number) in stacktraces
* `org.eclipse.jetty.util.log.stderr.SOURCE` (boolean, default "false") - used for showing of source (filename and line number) in stacktraces
* `org.eclipse.jetty.util.log.stderr.LONG` (boolean, default "false") - used for condensing the logger name package
* `org.eclipse.jetty.util.log.stderr.ESCAPE` (boolean, default "true") - used for escaping the output of the logger message.

Dropped properties

* `org.eclipse.jetty.util.log.class` - controlled by slf4j-api now
* `org.eclipse.jetty.util.log.announce` - controlled by slf4j-api now
* `org.eclipse.jetty.util.log.SOURCE` - we only have 1 logger now.

Renamed properties

* `org.eclipse.jetty.logging.IGNORED` (boolean, default "true") - used to control if Log.ignored() should be shown or not
* `org.eclipse.jetty.logging.THREAD_PADDING` (number, default "0") - used for minimum padding width of the Thread.name
* `org.eclipse.jetty.logging.SOURCE` (boolean, default "false") - used for showing of source (filename and line number) in stacktraces
* `org.eclipse.jetty.logging.NAME_CONDENSE` (boolean, default "true") - used for condensing the logger name package
* `org.eclipse.jetty.logging.MESSAGE_ESCAPE` (boolean, default "true") - used for escaping the output of the logger message

## Migrate source to use slf4j

Migrate code from ...

``` java
import org.eclipse.jetty.util.log.Log;
import org.eclipse.jetty.util.log.Logger;

private static final Logger LOG = Log.getLogger(DoSFilter.class);
```

... to ...

``` java
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

private static final Logger LOG = LoggerFactory.getLogger(DoSFilter.class);
```

The usage of `Log.ignore(Throwable)` would need to be migrated as well.

``` java
LOG.trace("IGNORED", cause);
```

The `org.eclipse.jetty.util.StacklessLogging` would be moved to
`org.eclipse.jetty.logging.StacklessLogging` (in the new `jetty-slf4j-impl` 
module) to provide the same behaviors as before for testing.

## New logging modules for jetty-home

The existing logging modules in Jetty 9.4.x are ...

* console-capture.mod
* jcl-slf4j.mod
* jul-impl.mod
* jul-slf4j.mod
* log4j2-api.mod
* log4j2-impl.mod
* log4j2-slf4j.mod
* log4j-impl.mod
* logback-impl.mod
* logging-jetty.mod
* logging-jul.mod
* logging-log4j.mod
* logging-log4j2.mod
* logging-logback.mod
* logging-slf4j.mod
* slf4j-api.mod
* slf4j-jul.mod
* slf4j-log4j.mod
* slf4j-log4j2.mod
* slf4j-logback.mod
* slf4j-simple-impl.mod

We would simplify this a great deal to just ...

* console-capture.mod
* logging-jetty.mod
* logging-jul.mod
* logging-log4j.mod
* logging-log4j2.mod
* logging-logback.mod

Where each logging module captures all logging implementations
via the various slf4j bridge jars, and enables the appropriate
slf4j implementation jar based on your logging module selection.

The new modules would also no longer force a forking of a JVM.

_Originally posted by @joakime in https://github.com/eclipse/jetty.project/issues/4567#issuecomment-585861202_

### PR DESCRIPTION
**Pull request by joakime**
Alternate approach from PR #4994

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

**Comment by gregw** [comment]
Closed in favour of #4994

**Comment by gregw** [comment]
Ooops wrong one

File: jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java [hunk 1]
```diff
@@ -32,7 +32,7 @@
      * Configuration keys specific to the StdErrAppender
      */
     static final String NAME_CONDENSE_KEY = "org.eclipse.jetty.logging.appender.NAME_CONDENSE";
-    static final String THREAD_PADDING_KEY = "org.eclipse.jetty.logging.appender.THREAD_PADDING";
+    static final String MESSAGE_ALIGN_KEY = "org.eclipse.jetty.logging.appender.MESSAGE_ALIGN";
```
Comment (gregw) [inline jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:35]: Much better name!  I thought TAG_PAD was some known name as it was in a TODO

File: jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java [hunk 2]
```diff
@@ -154,11 +154,15 @@ private void format(StringBuilder builder, JettyLogger logger, Level level, long
...
         builder.append(':');
-        builder.append(threadName); // TODO: support TAG_PAD configuration
+        builder.append(threadName);
         builder.append(':');
 
         // Message
-        builder.append(' ');
+        int padAmount = messageAlignColumn - builder.length();
+        if (padAmount > 0)
+            builder.append(" ".repeat(padAmount));
```
Comment (gregw) [inline jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: I'm just a little cautious that repeat might be less efficient than taking a substring of a longer string.

**Review by gregw — APPROVED** [review]
LGTM.    But perhaps check if substring of a string previously created by repeat is faster than repeat each time.  Probably down in the noise!

File: jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java (hunk 2 above)
Comment (joakime) [reply jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: Fair enough, i'll see what the numbers show.

I did look at the implementation of both in java.lang.String.

* String.substring() results in a new String object, based on "this.value" byte[] array, with an array copy to a new array for the new String object.
* String.repeat() results in a new String object, based on "this.value" byte[] array, with an array copy to a new array for the new String object.

Comment (joakime) [reply jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: @gregw here's the results of the `IndentBenchmark` (jmh) test.

https://github.com/eclipse/jetty.project/blob/jetty-10.0.x-4572-message-indent/tests/jetty-jmh/src/main/java/org/eclipse/jetty/logging/jmh/IndentBenchmark.java

```
Benchmark                                                               Mode  Cnt          Score           Error   Units
IndentBenchmark.testStringRepeatLarge                                  thrpt    7  186589212.883 ±  14318698.518   ops/s
IndentBenchmark.testStringRepeatLarge:·gc.alloc.rate                   thrpt    7       7822.988 ±       601.504  MB/sec
IndentBenchmark.testStringRepeatLarge:·gc.alloc.rate.norm              thrpt    7         88.000 ±         0.001    B/op
IndentBenchmark.testStringRepeatLarge:·gc.churn.G1_Eden_Space          thrpt    7       7854.853 ±      3677.899  MB/sec
IndentBenchmark.testStringRepeatLarge:·gc.churn.G1_Eden_Space.norm     thrpt    7         88.310 ±        40.367    B/op
IndentBenchmark.testStringRepeatLarge:·gc.count                        thrpt    7         18.000                  counts
IndentBenchmark.testStringRepeatLarge:·gc.time                         thrpt    7         64.000                      ms
IndentBenchmark.testStringRepeatSmall                                  thrpt    7  343002605.870 ±  25715711.800   ops/s
IndentBenchmark.testStringRepeatSmall:·gc.alloc.rate                   thrpt    7       9150.062 ±       688.673  MB/sec
IndentBenchmark.testStringRepeatSmall:·gc.alloc.rate.norm              thrpt    7         56.000 ±         0.001    B/op
IndentBenchmark.testStringRepeatSmall:·gc.churn.G1_Eden_Space          thrpt    7       9078.653 ±      4249.875  MB/sec
IndentBenchmark.testStringRepeatSmall:·gc.churn.G1_Eden_Space.norm     thrpt    7         55.546 ±        25.642    B/op
IndentBenchmark.testStringRepeatSmall:·gc.count                        thrpt    7         18.000                  counts
IndentBenchmark.testStringRepeatSmall:·gc.time                         thrpt    7         73.000                      ms
IndentBenchmark.testStringSubStringLarge                               thrpt    7  179315393.232 ± 109455137.398   ops/s
IndentBenchmark.testStringSubStringLarge:·gc.alloc.rate                thrpt    7       7518.461 ±      4593.109  MB/sec
IndentBenchmark.testStringSubStringLarge:·gc.alloc.rate.norm           thrpt    7         88.000 ±         0.001    B/op
IndentBenchmark.testStringSubStringLarge:·gc.churn.G1_Eden_Space       thrpt    7       7535.104 ±      6436.079  MB/sec
IndentBenchmark.testStringSubStringLarge:·gc.churn.G1_Eden_Space.norm  thrpt    7         87.144 ±        46.695    B/op
IndentBenchmark.testStringSubStringLarge:·gc.count                     thrpt    7         18.000                  counts
IndentBenchmark.testStringSubStringLarge:·gc.time                      thrpt    7         51.000                      ms
IndentBenchmark.testStringSubStringSmall                               thrpt    7  303492551.318 ±  41265934.036   ops/s
IndentBenchmark.testStringSubStringSmall:·gc.alloc.rate                thrpt    7       8096.412 ±      1102.379  MB/sec
IndentBenchmark.testStringSubStringSmall:·gc.alloc.rate.norm           thrpt    7         56.000 ±         0.001    B/op
IndentBenchmark.testStringSubStringSmall:·gc.churn.G1_Eden_Space       thrpt    7       8215.101 ±      3847.187  MB/sec
IndentBenchmark.testStringSubStringSmall:·gc.churn.G1_Eden_Space.norm  thrpt    7         56.618 ±        22.595    B/op
IndentBenchmark.testStringSubStringSmall:·gc.count                     thrpt    7         18.000                  counts
IndentBenchmark.testStringSubStringSmall:·gc.time                      thrpt    7         43.000                      ms
```

Comment (joakime) [reply jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: When I look at the results, I see that `String.repeat()` is slightly faster (in terms of ops/s.  4% faster in large cases, 12% faster in the small cases).
And the GC is about the same.
But with a higher error-bars on the substring version.

Comment (joakime) [reply jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: @gregw bump. could use a review of the benchmark results here.

Comment (gregw) [reply jetty-slf4j-impl/src/main/java/org/eclipse/jetty/logging/StdErrAppender.java:163]: repeat LGTM.    I think there may be other places we can do the same.... but another PR.

**Pull request merged**